### ISSUE TITLE
Conflict over ownership of file (napalm-nxos vs. napalm-iosxr)

### ISSUE THREAD

**Issue by he32** [napalm-automation/napalm-nxos#91]
### Description of Issue/Question

The napalm-nxos module wants to install a file which is also installed
by the napalm-iosxr module.  For us maintaining packages for these,
this is a big no-no.

### Did you follow the steps from https://github.com/napalm-automation/napalm#faq
[ X] Yes
[ ] No

### napalm-nxos version

```
0.5.2
```

### NX-OS version

```
Not relevant
```

### Steps to Reproduce the Issue
See that with python2.7, both packages want to install

`lib/python2.7/site-packages/test/unit/__init__.py`
and the corresponding `pyc` and `pyo` files.
That file name is, IMHO, way too generic,
well, the same can be said for all the files installed in the
`lib/python2.7/site-packages/test/` directory tree.

For now I've just put an ugly band-aid in place so all the `test/` files gets
omitted from the installation.

### Error Traceback

```
Not relevant.
```

**Comment by mirceaulinic** [comment]
Hi @he32 - `lib/python2.7/site-packages/test/unit/__init__.py` is not managed by any napalm library. 

Would you be able to do a `pip install napalm-iosxr` followed by `pip install napalm-nxos` or provide proper steps to reproduce this?

I just created a fresh venv and installed without seeing any issue:

```bash
admin@ip-172-31-11-15:~$ virtualenv iss91
New python executable in /home/admin/iss91/bin/python
Installing setuptools, pip, wheel...done.
admin@ip-172-31-11-15:~$ source iss91/bin/activate
(iss91) admin@ip-172-31-11-15:~$ pip install napalm-iosxr
Collecting napalm-iosxr
  Downloading napalm-iosxr-0.5.0.tar.gz
Collecting napalm-base>=0.18.0 (from napalm-iosxr)
  Downloading napalm-base-0.24.1.tar.gz
Collecting pyIOSXR>=0.50 (from napalm-iosxr)
  Downloading pyIOSXR-0.50.tar.gz
Collecting netaddr (from napalm-iosxr)
  Downloading netaddr-0.7.19-py2.py3-none-any.whl (1.6MB)
    100% |████████████████████████████████| 1.6MB 755kB/s
Collecting jtextfsm (from napalm-base>=0.18.0->napalm-iosxr)
  Downloading jtextfsm-0.3.1.tar.gz
Collecting jinja2 (from napalm-base>=0.18.0->napalm-iosxr)
  Downloading Jinja2-2.9.6-py2.py3-none-any.whl (340kB)
    100% |████████████████████████████████| 348kB 4.3MB/s
Collecting pyYAML (from napalm-base>=0.18.0->napalm-iosxr)
  Downloading PyYAML-3.12.tar.gz (253kB)
    100% |████████████████████████████████| 256kB 4.1MB/s
Collecting netmiko>=1.4.1 (from pyIOSXR>=0.50->napalm-iosxr)
  Downloading netmiko-1.4.1.tar.gz (44kB)
    100% |████████████████████████████████| 51kB 11.0MB/s
Collecting lxml>=3.2.4 (from pyIOSXR>=0.50->napalm-iosxr)
  Downloading lxml-3.8.0-cp27-cp27mu-manylinux1_x86_64.whl (6.8MB)
    100% |████████████████████████████████| 6.8MB 191kB/s
Collecting MarkupSafe>=0.23 (from jinja2->napalm-base>=0.18.0->napalm-iosxr)
  Downloading MarkupSafe-1.0.tar.gz
Collecting paramiko>=1.13.0 (from netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading paramiko-2.2.1-py2.py3-none-any.whl (176kB)
    100% |████████████████████████████████| 184kB 6.8MB/s
Collecting scp>=0.10.0 (from netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading scp-0.10.2-py2.py3-none-any.whl
Collecting pynacl>=1.0.1 (from paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading PyNaCl-1.1.2-cp27-cp27mu-manylinux1_x86_64.whl (539kB)
    100% |████████████████████████████████| 542kB 2.4MB/s
Collecting pyasn1>=0.1.7 (from paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading pyasn1-0.2.3-py2.py3-none-any.whl (53kB)
    100% |████████████████████████████████| 61kB 11.6MB/s
Collecting cryptography>=1.1 (from paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading cryptography-1.9.tar.gz (409kB)
    100% |████████████████████████████████| 419kB 3.2MB/s
Collecting bcrypt>=3.1.3 (from paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading bcrypt-3.1.3-cp27-cp27mu-manylinux1_x86_64.whl (57kB)
    100% |████████████████████████████████| 61kB 10.6MB/s
Collecting six (from pynacl>=1.0.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading six-1.10.0-py2.py3-none-any.whl
Collecting cffi>=1.4.1 (from pynacl>=1.0.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading cffi-1.10.0-cp27-cp27mu-manylinux1_x86_64.whl (392kB)
    100% |████████████████████████████████| 399kB 3.3MB/s
Collecting idna>=2.1 (from cryptography>=1.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading idna-2.5-py2.py3-none-any.whl (55kB)
    100% |████████████████████████████████| 61kB 11.1MB/s
Collecting asn1crypto>=0.21.0 (from cryptography>=1.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading asn1crypto-0.22.0-py2.py3-none-any.whl (97kB)
    100% |████████████████████████████████| 102kB 10.8MB/s
Collecting enum34 (from cryptography>=1.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading enum34-1.1.6-py2-none-any.whl
Collecting ipaddress (from cryptography>=1.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading ipaddress-1.0.18-py2-none-any.whl
Collecting pycparser (from cffi>=1.4.1->pynacl>=1.0.1->paramiko>=1.13.0->netmiko>=1.4.1->pyIOSXR>=0.50->napalm-iosxr)
  Downloading pycparser-2.17.tar.gz (231kB)
    100% |████████████████████████████████| 235kB 5.5MB/s
Building wheels for collected packages: napalm-iosxr, napalm-base, pyIOSXR, jtextfsm, pyYAML, netmiko, MarkupSafe, cryptography, pycparser
  Running setup.py bdist_wheel for napalm-iosxr ... done
  Stored in directory: /home/admin/.cache/pip/wheels/7a/94/54/5b667388a124074b736e913260ba06f55ab7022ad6bbc0cd85
  Running setup.py bdist_wheel for napalm-base ... done
  Stored in directory: /home/admin/.cache/pip/wheels/aa/5b/4b/22fd8e26d0304e77feaa84658c1e13bfb8cf041060babef122
  Running setup.py bdist_wheel for pyIOSXR ... done
  Stored in directory: /home/admin/.cache/pip/wheels/a0/6b/1a/87aed4784a29e37a2f708b398578e5d91fa4b516e3082982ad
  Running setup.py bdist_wheel for jtextfsm ... done
  Stored in directory: /home/admin/.cache/pip/wheels/6b/22/6b/b3d6deb6364579238b8017909729b5c1bad7fcd8331b4067c4
  Running setup.py bdist_wheel for pyYAML ... done
  Stored in directory: /home/admin/.cache/pip/wheels/2c/f7/79/13f3a12cd723892437c0cfbde1230ab4d82947ff7b3839a4fc
  Running setup.py bdist_wheel for netmiko ... done
  Stored in directory: /home/admin/.cache/pip/wheels/15/c7/f5/6421fe28f43b6cb92d5e0cc95a5f17dbb7d503ec9cbb6b4bd8
  Running setup.py bdist_wheel for MarkupSafe ... done
  Stored in directory: /home/admin/.cache/pip/wheels/88/a7/30/e39a54a87bcbe25308fa3ca64e8ddc75d9b3e5afa21ee32d57
  Running setup.py bdist_wheel for cryptography ... done
  Stored in directory: /home/admin/.cache/pip/wheels/ff/a5/ef/186bb4f6a89ef0bb8373bf53e5c9884b96722f0857bd3111b8
  Running setup.py bdist_wheel for pycparser ... done
  Stored in directory: /home/admin/.cache/pip/wheels/a8/0b/41/dc95621f9d3a0da7bc191b8a71f0e8182ffd3cc5f33ac55005
Successfully built napalm-iosxr napalm-base pyIOSXR jtextfsm pyYAML netmiko MarkupSafe cryptography pycparser
Installing collected packages: jtextfsm, MarkupSafe, jinja2, netaddr, pyYAML, napalm-base, six, pycparser, cffi, pynacl, pyasn1, idna, asn1crypto, enum34, ipaddress, cryptography, bcrypt, paramiko, scp, netmiko, lxml, pyIOSXR, napalm-iosxr
Successfully installed MarkupSafe-1.0 asn1crypto-0.22.0 bcrypt-3.1.3 cffi-1.10.0 cryptography-1.9 enum34-1.1.6 idna-2.5 ipaddress-1.0.18 jinja2-2.9.6 jtextfsm-0.3.1 lxml-3.8.0 napalm-base-0.24.1 napalm-iosxr-0.5.0 netaddr-0.7.19 netmiko-1.4.1 paramiko-2.2.1 pyIOSXR-0.50 pyYAML-3.12 pyasn1-0.2.3 pycparser-2.17 pynacl-1.1.2 scp-0.10.2 six-1.10.0
(iss91) admin@ip-172-31-11-15:~$ pip install napalm-nxos
Collecting napalm-nxos
  Downloading napalm-nxos-0.5.2.tar.gz
Requirement already satisfied: napalm-base>=0.18.0 in ./iss91/lib/python2.7/site-packages (from napalm-nxos)
Collecting pynxos (from napalm-nxos)
  Downloading pynxos-0.0.3.tar.gz
Requirement already satisfied: netaddr in ./iss91/lib/python2.7/site-packages (from napalm-nxos)
Collecting future (from napalm-nxos)
  Downloading future-0.16.0.tar.gz (824kB)
    100% |████████████████████████████████| 829kB 1.3MB/s
Requirement already satisfied: scp in ./iss91/lib/python2.7/site-packages (from napalm-nxos)
Requirement already satisfied: jtextfsm in ./iss91/lib/python2.7/site-packages (from napalm-base>=0.18.0->napalm-nxos)
Requirement already satisfied: jinja2 in ./iss91/lib/python2.7/site-packages (from napalm-base>=0.18.0->napalm-nxos)
Requirement already satisfied: pyYAML in ./iss91/lib/python2.7/site-packages (from napalm-base>=0.18.0->napalm-nxos)
Collecting requests>=2.7.0 (from pynxos->napalm-nxos)
  Downloading requests-2.18.1-py2.py3-none-any.whl (88kB)
    100% |████████████████████████████████| 92kB 5.4MB/s
Requirement already satisfied: paramiko in ./iss91/lib/python2.7/site-packages (from scp->napalm-nxos)
Requirement already satisfied: MarkupSafe>=0.23 in ./iss91/lib/python2.7/site-packages (from jinja2->napalm-base>=0.18.0->napalm-nxos)
Collecting urllib3<1.22,>=1.21.1 (from requests>=2.7.0->pynxos->napalm-nxos)
  Downloading urllib3-1.21.1-py2.py3-none-any.whl (131kB)
    100% |████████████████████████████████| 133kB 4.1MB/s
Requirement already satisfied: idna<2.6,>=2.5 in ./iss91/lib/python2.7/site-packages (from requests>=2.7.0->pynxos->napalm-nxos)
Collecting certifi>=2017.4.17 (from requests>=2.7.0->pynxos->napalm-nxos)
  Downloading certifi-2017.4.17-py2.py3-none-any.whl (375kB)
    100% |████████████████████████████████| 378kB 2.7MB/s
Collecting chardet<3.1.0,>=3.0.2 (from requests>=2.7.0->pynxos->napalm-nxos)
  Downloading chardet-3.0.4-py2.py3-none-any.whl (133kB)
    100% |████████████████████████████████| 143kB 5.2MB/s
Requirement already satisfied: pynacl>=1.0.1 in ./iss91/lib/python2.7/site-packages (from paramiko->scp->napalm-nxos)
Requirement already satisfied: pyasn1>=0.1.7 in ./iss91/lib/python2.7/site-packages (from paramiko->scp->napalm-nxos)
Requirement already satisfied: cryptography>=1.1 in ./iss91/lib/python2.7/site-packages (from paramiko->scp->napalm-nxos)
Requirement already satisfied: bcrypt>=3.1.3 in ./iss91/lib/python2.7/site-packages (from paramiko->scp->napalm-nxos)
Requirement already satisfied: six in ./iss91/lib/python2.7/site-packages (from pynacl>=1.0.1->paramiko->scp->napalm-nxos)
Requirement already satisfied: cffi>=1.4.1 in ./iss91/lib/python2.7/site-packages (from pynacl>=1.0.1->paramiko->scp->napalm-nxos)
Requirement already satisfied: asn1crypto>=0.21.0 in ./iss91/lib/python2.7/site-packages (from cryptography>=1.1->paramiko->scp->napalm-nxos)
Requirement already satisfied: ipaddress in ./iss91/lib/python2.7/site-packages (from cryptography>=1.1->paramiko->scp->napalm-nxos)
Requirement already satisfied: enum34 in ./iss91/lib/python2.7/site-packages (from cryptography>=1.1->paramiko->scp->napalm-nxos)
Requirement already satisfied: pycparser in ./iss91/lib/python2.7/site-packages (from cffi>=1.4.1->pynacl>=1.0.1->paramiko->scp->napalm-nxos)
Building wheels for collected packages: napalm-nxos, pynxos, future
  Running setup.py bdist_wheel for napalm-nxos ... done
  Stored in directory: /home/admin/.cache/pip/wheels/d3/81/00/5e0b679719ca4b974627c6b40108c1b1b28abb2bddddab981b
  Running setup.py bdist_wheel for pynxos ... done
  Stored in directory: /home/admin/.cache/pip/wheels/84/ff/be/f0989593202f198e5872d1258df4956d9f6646d72fe22c2da1
  Running setup.py bdist_wheel for future ... done
  Stored in directory: /home/admin/.cache/pip/wheels/c2/50/7c/0d83b4baac4f63ff7a765bd16390d2ab43c93587fac9d6017a
Successfully built napalm-nxos pynxos future
Installing collected packages: urllib3, certifi, chardet, requests, future, pynxos, napalm-nxos
Successfully installed certifi-2017.4.17 chardet-3.0.4 future-0.16.0 napalm-nxos-0.5.2 pynxos-0.0.3 requests-2.18.1 urllib3-1.21.1
(iss91) admin@ip-172-31-11-15:~$
```

**Comment by dbarrosop** [comment]
@he32, this is an opensource project driven mostly by people that does this to help the community. Instead of big claims and complaints it would be much better to be polite and work together to identify the issue (if any) and correct it.

To the issue at hand, test files are not installed by pip so as mircea suggested, please, provide steps to reproduce the issue so we can identify what's going on.

**Comment by he32** [comment]
> Hi @he32 - `lib/python2.7/site-packages/test/unit/__init__.py`
> is not managed by any napalm library.
>
> Would you be able to do a `pip install nalam-iosxr` followed by
> `pip install napalm-nxos` or provide proper steps to reproduce
> this?

Hmm...  Since this is pkgsrc (pkgsrc-wip, actually), ref. mirror at

  https://github.com/NetBSD/pkgsrc-wip

we're using the pkgsrc packaging tools to build and install this
software, and if I understand correctly, "pip" is then an
antithesis of this method.

The pkgsrc build infrastructure basically does

```
do-install:
        (cd ${WRKSRC}/${PYSETUPSUBDIR} && ${SETENV} ${INSTALL_ENV} ${MAKE_ENV} \
         ${PYTHONBIN} ${PYSETUP} ${PYSETUPARGS} "install" ${_PYSETUPINSTALLARGS})
```

where among others:

```
PYSETUP?=  setup.py
PYSETUPINSTALLARGS?=    #empty
PYSETUPOPTARGS?=        -c -O1
_PYSETUPINSTALLARGS=    ${PYSETUPINSTALLARGS} ${PYSETUPOPTARGS} ${_PYSETUPTOOLSINSTALLARGS}
_PYSETUPINSTALLARGS+=   --root=${DESTDIR:Q}
```

and save from environment setup, ends up doing

```
/usr/pkg/bin/python2.7 setup.py  "install"  -c -O1 --root=/usr/pkgsrc/wip/py-napalm-nxos/work/.destdir
```

Done this way, all the files under test/ in the distribution
somehow ends up under python's site-packages/ directory, and I
need then to add a rather ugly

```
post-install:
        rm -rf ${DESTDIR}/${PREFIX}/${PYSITELIB}/test/
```

fix to the (py-)napalm-nxos and (py-)napalm-iosxr Makefiles.

I'll readily admit that I'm no expert in python...

Hm, I see the test/ directory files end up in
`work/napalm-nxos-0.5.2/napalm_nxos.egg-info/SOURCES.txt`
presumably after the "build" phase.  And the build phase
basically does `/usr/pkg/bin/python2.7 setup.py build` and prints
among others

```
writing manifest file 'napalm_nxos.egg-info/SOURCES.txt'
reading manifest file 'napalm_nxos.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'napalm_nxos.egg-info/SOURCES.txt'
```

and sure enough, these entries end up in `SOURCES.txt`:

```
test/__init__.py
test/unit/TestDriver.py
test/unit/__init__.py
test/unit/conftest.py
test/unit/test_getters.py
```

Clues welcome.

Regards,

Håvard

**Comment by dbarrosop** [comment]
I can't speak for the rest but I have never used that tool so I am not sure how it works or how to fix it but if you have a solution we will happy review it and apply it. The way I usually package python packages myself is by using fpm to package an entire virtualenv that had the libraries installed with pip so I haven't encountered the problem you mention.

**Comment by ktbyers** [comment]
This looks like there might be an issue with the NAPALM `setup.py`. I will see if I can look at it.

Assigning issue over to me.

**Comment by he32** [comment]
> This looks like there might be an issue with the NAPALM
> `setup.py`. I will see if I can look at it.
>
> Assigning issue over to me.

Ah, yes...

It looks like

-  packages=find_packages(),
+  packages=find_packages(exclude=["test", "test.*"]),

is an appropriate change.

Regards,

Håvard
